File: src/main/java/ro/nicolaemariusghergu/queryit/service/impl/CategoryServiceImpl.java
package ro.nicolaemariusghergu.queryit.service.impl;

import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;
import ro.nicolaemariusghergu.queryit.model.Category;
import ro.nicolaemariusghergu.queryit.repository.CategoryRepository;
import ro.nicolaemariusghergu.queryit.service.CategoryService;

import java.util.List;
import java.util.Optional;

@Service
public record CategoryServiceImpl(CategoryRepository categoryRepository) implements CategoryService {

    @NonNull
    @Override
    public Optional<Category> findById(@NonNull Long id) {
        return categoryRepository.findById(id);
    }

    @NonNull
    @Override
    public List<Category> findAll() {
        return categoryRepository.findAll();
    }

    @Override
    public Optional<Category> findByName(String name) {
        return categoryRepository.findByName(name);
    }

    @Override
    public <S extends Category> S save(S entity) {
        return categoryRepository.save(entity);
    }

    @Override
    public <S extends Category> List<S> saveAll(Iterable<S> entities) {
        return categoryRepository.saveAll(entities);
    }

    @NonNull
    @Override
    public void deleteById(@NonNull Long id) {
        categoryRepository.deleteById(id);
    }
}

File: src/main/kotlin/ro/nicolaemariusghergu/queryit/service/impl/CategoryServiceImpl.kt
package ro.nicolaemariusghergu.queryit.service.impl

import lombok.extern.slf4j.Slf4j
import org.springframework.http.ResponseEntity
import org.springframework.lang.NonNull
import org.springframework.stereotype.Service
import ro.nicolaemariusghergu.queryit.dto.CategoryDto
import ro.nicolaemariusghergu.queryit.mapper.CategoryMapper
import ro.nicolaemariusghergu.queryit.model.*
import ro.nicolaemariusghergu.queryit.repository.CategoryRepository
import ro.nicolaemariusghergu.queryit.service.CategoryService
import java.util.function.Function
import java.util.function.Supplier

@Service
class CategoryServiceImpl(private val categoryRepository: CategoryRepository) : CategoryService {

    override fun getCategories(): ResponseEntity<MutableList<CategoryDto>> {
        return ResponseEntity.ok(categoryRepository.findAll().stream()
                .map { category: Category -> CategoryMapper.INSTANCE.categoryToCategoryDto(category) }
                .toList())
    }

    override fun addCategory(categoryDto: CategoryDto): ResponseEntity<Long> {
        categoryRepository.save(CategoryMapper.INSTANCE
                .categoryDtoToCategory(categoryDto))
        return ResponseEntity.ok(categoryDto.id)
    }
}